diff -Nur a/local/proxy.py b/local/proxy.py
--- a/local/proxy.py	2014-09-27 00:30:50.299673000 +0800
+++ b/local/proxy.py	2014-09-27 00:31:17.799672000 +0800
@@ -50,10 +50,11 @@
 import os
 import sys
 import sysconfig
+import glob
 
 reload(sys).setdefaultencoding('UTF-8')
 sys.dont_write_bytecode = True
-sys.path = [(os.path.dirname(__file__) or '.') + '/packages.egg/noarch'] + sys.path + [(os.path.dirname(__file__) or '.') + '/packages.egg/' + sysconfig.get_platform().split('-')[0]]
+sys.path += glob.glob('%s/*.egg' % os.path.dirname(os.path.abspath(__file__)))
 
 try:
     __import__('gevent.monkey', fromlist=['.']).patch_all()
